### PR TITLE
feat: Trigger updateArtifacts for update-lockfile rangeStrategy with no packageFile content updated

### LINKED ISSUE #3521: Support rangeStrategy="update-lockfile" for Composer
See [rangeStrategy docs](https://renovatebot.com/docs/configuration-options/#rangestrategy).

This currently works for npm and yarn only. Composer would be a good next manager to support.

To reproduce the problem/feature, you'd want:

1. `composer.json` containing a dependency with a range (e.g. `^1.0.0`)
2. `composer.lock` locking that dependency to a specific version (e.g. `v1.0.1`)
3. A new update for that dependency to be available (e.g. `v1.1.0`)

Currently, Renovate won't do anything in the above example because `v1.1.0` is "in range" with `^1.0.0`. What we want is to give the option to define a "update-lockfile" rangeStrategy and have Renovate raise a PR updating the `composer.lock` only (i.e. to `v1.1.0` in the above example).

It's possible this already "half works" because many of the pieces are together. Searching for the strings `update-lockfile` and `lockfileUpdate` should identify the majority of places where the logic is currently implemented (i.e. for npm/yarn).

However it's also possible we have some technical debt to refactor out - this might be one of the last areas with some npm-specific logic at the lib/workers/* abstraction layer of Renovate.

### PR DESCRIPTION
**Pull request by pvgnd**
When an upgrade does not produce any changes in the package file, we never run updateArtifacts.  As a consequence, the scenario described in #3521 don't work.  

Since this MR trigger the updateArtifacts for update-lockfile rangeStrategy when packageFile has no change, the composer.lock file is updated.

Closes #3521

### CODE REVIEW

**Review by viceice — DISMISSED** [review]

**Comment by viceice** [comment]
This looks more clean to me. Maybe we should add some tests?

**Comment by pvgnd** [comment]
of course, I will

**Comment by pvgnd** [comment]
@viceice I just added a test

**Review by viceice — COMMENTED** [review]
Have you done some real tests? Can we see some samples?

**Comment by pvgnd** [comment]
Yes,

here is a replace test : https://github.com/pvgnd-testing/renovate-composer-replace/pull/2/files  
here is a update-lockfile test : https://github.com/pvgnd-testing/renovate-composer-update-lockfile/pull/2/files

**Review by viceice — DISMISSED** [review]

**Comment by rarkins** [comment]
Is Composer the only manager affected by this?

**Comment by pvgnd** [comment]
No, all managers that are compatible with update-lockfile strategy. But tests were still green so I guess no impact.  
The fact is that we need to updateArtifacts even if the packageFile content has not been update with the update-lockfile strategy. So should be OK for other manager as well. But maybe we should have something to check that a manager allowed the update-lockfile strategy ?  

 As I don't know the project code very well, I'd be happy to have some help or input on the subject.

**Comment by rarkins** [comment]
Can you test it out with a npm or yarn repo to verify it still works? Those were the initial `update-lockfile` implementations and a little hacked in. We need to make sure this solution is compatible with it.

**Comment by pvgnd** [comment]
Here is the yarn test:  
- replace strategy: https://github.com/pvgnd-testing/renovate-yarn-replace/pull/1
- update-lockfile strategy: https://github.com/pvgnd-testing/renovate-yarn-update-lockfile/pull/1

The results seems weird to me but I don't know yarn very well. By the way the result is the same with the master version of renovate.

Here is the composer test:
- replace strategy: https://github.com/pvgnd-testing/renovate-composer-replace/pull/2/files
- update-lockfile strategy: https://github.com/pvgnd-testing/renovate-composer-update-lockfile/pull/2/files

**Comment by rarkins** [comment]
Unless I'm completely confused, the update-lockfile strategy should not result in the `package.json` changing unless necessary. But in https://github.com/pvgnd-testing/renovate-yarn-update-lockfile/pull/1 it is:

image

**Comment by pvgnd** [comment]
I agree, but if you run the master version of renovate against the same repository, it will bump the package.json also. So, seems to be an existing bug in this specific manager.

File: lib/workers/branch/get-updated.ts [hunk 1]
```diff
@@ -62,6 +62,9 @@ export async function getUpdatedPackageFiles(
         if (res) {
           if (res === existingContent) {
             logger.debug({ packageFile, depName }, 'No content changed');
+            if (upgrade.rangeStrategy === 'update-lockfile') {
+              updatedFileContents[packageFile] = res;
```
Comment (rarkins) [inline lib/workers/branch/get-updated.ts:66]:
```suggestion
              logger.debug({ packageFile, depName }, 'update-lockfile add');
              updatedFileContents[packageFile] = res;
```

**Comment by rarkins** [comment]
Needs coverage fix too

**Comment by monojp** [comment]
Hello, I was looking into renovate to automate Composer updates and apparently this PR would improve it even more. Is this still in progress, can somebody else help out here?

**Comment by rarkins** [comment]
@monojp if you can demonstrate this feature working on a public composer repo, that would help too, because that's also needed before we can merge.

**Review by viceice — COMMENTED** [review]
LGTM, but need real test runs. Do we have a test repo for this?

**Review by viceice — APPROVED** [review]
OK, found it, seems to work for that use case

**Comment by viceice** [comment]
Tests:
https://github.com/viceice-tests/renovate-composer-replace/pull/1
https://github.com/viceice-tests/renovate-yarn-update-lockfile2/pull/2
https://github.com/viceice-tests/renovate-yarn-update-lockfile2/pull/1

**Comment by viceice** [comment]
https://github.com/viceice-tests/renovate-yarn-update-lockfile/pulls

**Comment by monojp** [comment]
Wow that's really great. Thank you very much @rarkins and @viceice!

**Pull request merged**